### ISSUE TITLE
LightGBM can't learn simple rule

### ISSUE THREAD

**Issue by gendrelom** [microsoft/LightGBM#5493]
## Description
I'm trying to use LightGBM for binary classification. I created a synthetic dataset which has 6 features and one label with two classes: True or False. I randomly create features and then apply the rule - if the features pass through the filter, then the label is more likely to be true. 
It's very simple to manually apply this filter to the dataset and get precision ~0.8. However, LightGBM can't find this rule to achieve this precision. Results worse by 3-5 times. 
Also, I don't need to have good accuracy, I need precision greater than or equal to 0.75 and examples which passes through the filter(107 examples in this case).

So, what can I do to achieve this?

## Reproducible example
```
import pandas as pd
import numpy as np
import tqdm
import sklearn
import lightgbm as lgb

rows_len = 4000

df_arr = []
for i in range(rows_len):
  f1 = np.random.randint(0, 11)
  f2 = np.random.randint(0, 11)
  f3 = np.random.randint(0, 11)
  f4 = np.random.randint(0, 11)
  f5 = np.random.randint(0, 11)
  f6 = np.random.randint(0, 11)

  y = False
  if f1 > 2 and f2 < 5 and (f3 > 2 and f3 < 9) and (f4 > 1 and f4 < 7) and (f5 > 3 and f5 < 10) and (f6 > 3 and f6 < 10):
    y = True if np.random.randint(1,11) > 2 else False
  else:
    y = True if np.random.randint(1,11) > 5 else False

  df_arr.append([y, f1, f2, f3, f4, f5, f6])


df = pd.DataFrame.from_records(df_arr, columns=['y', 'f1', 'f2', 'f3', 'f4', 'f5', 'f6'])
df['y'] = df['y'].astype('bool')
df['f1'] = df['f1'].astype('int')
df['f2'] = df['f2'].astype('int')
df['f3'] = df['f3'].astype('int')
df['f4'] = df['f4'].astype('int')
df['f5'] = df['f5'].astype('int')
df['f6'] = df['f6'].astype('int')
x_columns = ['f1', 'f2', 'f3', 'f4', 'f5', 'f6']

print(df['y'].sum()) # 1977

true_pos_and_neg = df[(df.f1 > 2) & (df.f2 < 5) & (df.f3 > 2) & (df.f3 < 9) & (df.f4 > 1) & (df.f4 < 7) & (df.f5 > 3) & (df.f5 < 10) & (df.f6 > 3) & (df.f6 < 10)]
print(len(true_pos_and_neg)) # 107

true_pos_len = true_pos_and_neg.y.sum()
print(true_pos_len) # 86

print(true_pos_len / len(true_pos_and_neg)) # 0.8037383177570093

fold_size = int(rows_len / 5)

k_folds = []
for i in range(int(len(df)/fold_size)):

  ind = i*fold_size

  train_df = df[:ind].append(df[ind+fold_size:]).reset_index(drop=True)
  test_df = df[ind:ind+fold_size]

  k_folds.append({
      'train_df': train_df.reset_index(drop=True),
      'test_df': test_df.reset_index(drop=True)
  })

lgbm_param_grid = {
    'num_threads': [4],
    'reg_alpha': [3, 1, 0.5, 0],
    'reg_lambda': [3, 1, 0.5, 0],
    'max_depth': [-1, 10, 30, ],
    'learning_rate': [0.9, 0.5, 0.1],
    'n_estimators': [3, 5, 10, 30],
    'num_leaves': [50, 70, 100],
    }

precision_threshold = 0.75

model = lgb.LGBMClassifier()
best_score = 0
best_precision = 0
best_retrieved = 0

for g in tqdm.tqdm(sklearn.model_selection.ParameterGrid(lgbm_param_grid)):
    model.set_params(**g)

    probs = np.array([])
    labels = np.array([])

    for fold in k_folds:
      train_df = fold['train_df']
      test_df = fold['test_df'] 

      model.fit(train_df[x_columns], train_df['y'])

      probs = np.append(probs, model.predict_proba(test_df[x_columns])[:, 1])
      labels = np.append(labels, test_df['y'])


    best_threshold_precision = 0
    best_threshold_retrieved = 0
    best_threshold = 0

    for threshold in np.arange(0.5, 1, 0.05):
      probs_with_threshold = probs > threshold

      true_pos = 0
      true_pos_and_neg = 0

      for i in range(len(labels)):
        if labels[i] and probs_with_threshold[i]:
          true_pos += 1 
          true_pos_and_neg += 1
        elif probs_with_threshold[i] and not labels[i]:
          true_pos_and_neg += 1

      precision = 0 if true_pos_and_neg == 0 else true_pos / true_pos_and_neg
      retrieved = probs_with_threshold.sum()

      if precision >= precision_threshold and (precision * retrieved) > (best_threshold_precision * best_threshold_retrieved):
        best_threshold_precision = precision
        best_threshold_retrieved = retrieved
        best_threshold = threshold

    if best_threshold_precision >= precision_threshold and best_threshold_precision * best_threshold_retrieved >= best_score:
        best_score = best_threshold_precision * best_threshold_retrieved
        best_precision = best_threshold_precision
        best_grid = g
        best_retrieved = best_threshold_retrieved
    
print(best_precision, best_retrieved) # 0.7714285714285715, 35
print(best_grid)
# {'learning_rate': 0.9,
# 'max_depth': 10,
# 'n_estimators': 10,
# 'num_leaves': 100,
# 'num_threads': 4,
# 'reg_alpha': 0.5,
# 'reg_lambda': 1}
```

## Environment info

LightGBM version or commit hash: Version: 2.2.3

Command(s) you used to install LightGBM: preinstalled version on Colab

**Comment by jameslamb** [comment]
Thanks for using LightGBM!

I haven't tried your example yet, but I suspect that I know how to fix the problem you're having.

LightGBM has some parameters used to control overfitting, and these have default values which are set to hopefully work well with medium-sized datasets (greater than 10,000 rows). When working with very very small datasets like the one in your example (107 rows), those parameters need to be adjusted if you want LightGBM to overfit to the provided data.

Please try adding the following in `params`:

* `min_data_in_bin=1` ([documentation link](https://lightgbm.readthedocs.io/en/v3.3.2/Parameters.html#min_data_in_leaf))
* `min_data_in_leaf=1` ([documentation link](https://lightgbm.readthedocs.io/en/v3.3.2/Parameters.html#min_data_in_leaf))

For more details on this, see these related Stack Overflow answers:

* ["quick lightgbm vs xgboost check. lightgbm makes the wrong cut?"](https://stackoverflow.com/questions/68733851/quick-lightgbm-vs-xgboost-check-lightgbm-makes-the-wrong-cut/68736364#68736364)
* ["LightGBM depth not increasing from 0"](https://stackoverflow.com/questions/73392144/lightgbm-depth-not-increasing-from-zero/73396385#73396385)

**Comment by gendrelom** [comment]
@jameslamb I tried to use `LGBMClassifier` with 40k rows(instead of 4k), script generated 959 `true_pos_and_neg` and 761 `true_pos`, so for manual filtering precision = 0.79.
I tried to optimize hyperparameters and easily got 992 `true_pos_and_neg` with precision = 0.77(it's 764 true positive) using `LGBMClassifier`.
So, it's seems that problem is the dataset size, but I can't increase it yet. 

What hyperparameters should I use for the dataset in my first post?
